File: backend/api/face_matcher.py
import numpy as np


def cosine_similarity(a, b):
    a = np.array(a)
    b = np.array(b)

    return np.dot(a, b) / (
        np.linalg.norm(a)
        * np.linalg.norm(b)
    )


def match_faces(
    stored_embedding,
    current_embedding,
    threshold=0.65
):
    score = cosine_similarity(
        stored_embedding,
        current_embedding
    )

    return score >= threshold, float(score)

File: mobile-app/src/ai/FaceMatcher.ts

export function cosineSimilarity(
  a: number[],
  b: number[]
) {
  let dot = 0;
  let magA = 0;
  let magB = 0;

  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    magA += a[i] * a[i];
    magB += b[i] * b[i];
  }

  return dot / (
    Math.sqrt(magA) *
    Math.sqrt(magB)
  );
}

export function matchFaces(
  registered: number[],
  current: number[]
) {
  const score = cosineSimilarity(
    registered,
    current
  );

  console.log("🎯 Similarity:", score);

  const matched = score > 0.7;

  console.log(
    matched
      ? "✅ AUTHENTICATED"
      : "❌ ACCESS DENIED"
  );

  return matched;
}